src/threading.c

```diff
@@ -408,16 +408,24 @@ jl_ptls_t jl_init_threadtls(int16_t tid)
 
 JL_DLLEXPORT jl_gcframe_t **jl_adopt_thread(void) JL_NOTSAFEPOINT_LEAVE
 {
+    // jl_init_threadtls puts us in a GC unsafe region, so ensure GC isn't running.
+    // we can't use a normal safepoint because we don't have signal handlers yet.
+    // we also can't use jl_safepoint_wait_gc because that assumes we're in a task.
+    while (jl_atomic_load_relaxed(&jl_gc_running) || jl_atomic_load_acquire(&jl_gc_running)) {
+        jl_cpu_pause();
+    }
+    jl_atomic_fetch_add(&jl_gc_disable_counter, 1);
```

Comment:
Can you change the order here?

1. Request GC is off
2. Wait for any GC to finish

There is still a window I am worried about. In `jl_gc_collect` we:

1. Check if GC is allowed to run
2. Win the race to become GC thread and mark gc_running

So there is a possible interleaving.

Where 
1. Check GC is allowed to run
2. Foreign Thread requests GC off
3. Foreign Thread waits for GC to finish
4. Win the race to become GC thread and mark gc_running

Reply:
Thinking about it some more we will need to check disable_counter after winning the race